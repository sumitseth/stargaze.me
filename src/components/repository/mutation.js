import gql from 'graphql-tag';

export const addStarMutation = gql`
  mutation($starrableId: ID!) {
    addStar(input: { starrableId: $starrableId }) {
      starrable {
        id
      }
    }
  }
`;

export const removeStarMutation = gql`
  mutation($starrableId: ID!) {
    removeStar(input: { starrableId: $starrableId }) {
      starrable {
        id
      }
    }
  }
`;

export const addStarAction = id => ({
  variables: { starrableId: id },
  update: proxy => {
    proxy.writeFragment({
      id: `Repository:${id}`,
      fragment: gql`
        fragment repository on Repository {
          viewerHasStarred
        }
      `,
      data: { viewerHasStarred: true, __typename: 'Repository' },
    });
  },
});

export const removeStarAction = id => ({
  variables: { starrableId: id },
  optimisticResponse: {
    __typename: 'Mutation',
    removeStar: {
      starrable: {
        id: id,
        __typename: 'Repository',
      },
    },
  },
  update: proxy => {
    proxy.writeFragment({
      id: `Repository:${id}`,
      fragment: gql`
        fragment repository on Repository {
          viewerHasStarred
        }
      `,
      data: { viewerHasStarred: false, __typename: 'Repository' },
    });
  },
});
