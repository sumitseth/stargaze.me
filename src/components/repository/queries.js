import gql from 'graphql-tag';

export const searchReposQuery = gql`
  query($query: String!, $after: String) {
    search(type: REPOSITORY, query: $query, first: 25, after: $after) {
      repositoryCount
      pageInfo {
        endCursor
        hasNextPage
      }
      nodes {
        ... on Repository {
          id
          name
          url
          description
          viewerHasStarred
        }
      }
    }
  }
`;

export const starredReposQuery = gql`
  query StarredReposQuery($numRepos: Int, $after: String) {
    viewer {
      id
      name
      avatarUrl
      starredRepositories(first: $numRepos, after: $after) {
        totalCount
        pageInfo {
          endCursor
          hasNextPage
        }
        nodes {
          id
          name
          description
          url
          viewerHasStarred
        }
      }
    }
  }
`;
