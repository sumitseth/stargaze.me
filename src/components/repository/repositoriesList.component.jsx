import React from 'react';
import RepositoryItemComponent from './repositoryItem.component';
import { repositoriesListTestSelectors } from './repository.test.selectors';

export const RepositoriesList = ({ repositories }) => {
  return (
    <ul data-testid={repositoriesListTestSelectors.list}>
      {repositories.nodes.map(({ name, description, url, id }) => (
        <RepositoryItemComponent
          name={name}
          description={description}
          url={url}
          key={id}
        />
      ))}
    </ul>
  );
};

export default RepositoriesList;
