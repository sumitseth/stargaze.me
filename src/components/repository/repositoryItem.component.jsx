import React from 'react';
import PropTypes from 'prop-types';
import { repositoryTestSelectors } from './repository.test.selectors';

const RepositoryItem = ({ name, description, url }) => (
  <div>
    <h2 data-testid={repositoryTestSelectors.name}>{name}</h2>
    <a
      data-testid={repositoryTestSelectors.url}
      href={url}
      target="_blank"
      rel="noopener noreferrer"
    >
      {url}
    </a>
    <p data-testid={repositoryTestSelectors.description}>{description}</p>
  </div>
);

RepositoryItem.propTypes = {
  name: PropTypes.string.isRequired,
  url: PropTypes.string.isRequired,
  description: PropTypes.string,
};

export default RepositoryItem;
