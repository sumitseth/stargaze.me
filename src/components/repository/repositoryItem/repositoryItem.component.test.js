import { render, getByTestId } from '@testing-library/react';
import renderer from 'react-test-renderer';
import React from 'react';
import RepositoryItem from './repositoryItem.component';
import { repositoryTestSelectors } from './repository.testSelectors';

const repository = {
  name: 'stargaze.me',
  url: 'http://stargaze.me',
  description: 'star',
  viewerHasStarred: true,
};

it('renders without crashing', () => {
  expect.assertions(3);
  const { container } = render(<RepositoryItem {...repository} />);
  const name = getByTestId(container, repositoryTestSelectors.name);
  const description = getByTestId(
    container,
    repositoryTestSelectors.description
  );
  const url = getByTestId(container, repositoryTestSelectors.url);

  expect(name.textContent).toBe(repository.name);
  expect(description.textContent).toBe(repository.description);
  expect(url.textContent).toBe(repository.url);
});

it('matches snapshot', () => {
  const tree = renderer.create(<RepositoryItem {...repository} />).toJSON();
  expect(tree).toMatchSnapshot();
});
