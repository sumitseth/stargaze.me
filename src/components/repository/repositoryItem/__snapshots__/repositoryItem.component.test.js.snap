// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div>
  <h2
    data-testid="repository-name"
  >
    stargaze.me
  </h2>
  <a
    data-testid="repository-url"
    href="http://stargaze.me"
    rel="noopener noreferrer"
    target="_blank"
  >
    http://stargaze.me
  </a>
  <img
    alt="viewerHasStarred"
    onClick={[Function]}
    src={
      Object {
        "0": "s",
        "1": "t",
        "10": "e",
        "11": ".",
        "12": "p",
        "13": "n",
        "14": "g",
        "2": "a",
        "3": "r",
        "4": "-",
        "5": "a",
        "6": "c",
        "7": "t",
        "8": "i",
        "9": "v",
        "default": "star-active.png",
      }
    }
    style={
      Object {
        "cursor": "pointer",
        "verticalAlign": "middle",
      }
    }
  />
  <p
    data-testid="repository-description"
  >
    star
  </p>
</div>
`;
